hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -423,56 +423,54 @@ public boolean containsEntityHolder(EntityKey key) {
 
 	@Override
 	public void postLoad(JdbcValuesSourceProcessingState processingState, Consumer<EntityHolder> holderConsumer) {
-		if ( entitiesByKey == null ) {
+		if ( CollectionHelper.isEmpty( processingState.getLoadingEntityHolders() ) ) {
 			return;
 		}
 		final Callback callback = processingState.getExecutionContext().getCallback();
 		final EventListenerGroup<PostLoadEventListener> listenerGroup = getSession().getFactory()
 				.getFastSessionServices()
 				.eventListenerGroup_POST_LOAD;
-		final PostLoadEvent postLoadEvent = getSession().isEventSource() ? new PostLoadEvent( getSession().asEventSource() ) : null;
-		for ( Iterator<EntityHolderImpl> iterator = entitiesByKey.values().iterator(); iterator.hasNext(); ) {
-			final EntityHolderImpl holder = iterator.next();
-			if ( holder.processingState == processingState ) {
-				if ( holderConsumer != null ) {
-					holderConsumer.accept( holder );
-				}
-				if ( holder.entity == null ) {
-					// It's possible that we tried to load an entity and found out it doesn't exist,
-					// in which case we added an entry with a null proxy and entity.
-					// Remove that empty entry on post load to avoid unwanted side effects
-					iterator.remove();
-					continue;
-				}
-				if ( postLoadEvent != null ) {
-					postLoadEvent.reset();
-					postLoadEvent.setEntity( holder.entity )
-							.setId( holder.entityKey.getIdentifier() )
-							.setPersister( holder.getDescriptor() );
-					listenerGroup.fireEventOnEachListener(
-							postLoadEvent,
-							PostLoadEventListener::onPostLoad
-					);
-				}
+		final PostLoadEvent postLoadEvent = processingState.getPostLoadEvent();
+		final Iterator<EntityHolder> iterator = processingState.getLoadingEntityHolders().iterator();
+		while ( iterator.hasNext() ) {
+			final EntityHolder holder = iterator.next();
+			if ( holderConsumer != null ) {
+				holderConsumer.accept( holder );
+			}
+			if ( holder.getEntity() == null ) {
+				// It's possible that we tried to load an entity and found out it doesn't exist,
+				// in which case we added an entry with a null proxy and entity.
+				// Remove that empty entry on post load to avoid unwanted side effects
+				entitiesByKey.remove( holder.getEntityKey() );
+				continue;
+			}
+			if ( postLoadEvent != null ) {
+				postLoadEvent.reset();
+				postLoadEvent.setEntity( holder.getEntity() )
+						.setId( holder.getEntityKey().getIdentifier() )
+						.setPersister( holder.getDescriptor() );
+				listenerGroup.fireEventOnEachListener(
+						postLoadEvent,
+						PostLoadEventListener::onPostLoad
+				);
+			}
 
-				if ( callback != null ) {
+			if ( callback != null ) {
+				callback.invokeAfterLoadActions(
+						holder.getEntity(),
+						holder.getDescriptor(),
+						getSession()
+				);
+				if ( holder.isReloaded() ) {
```

Comment:
I actually think this branch can be removed. We don't need the `reloaded` flag anymore.

Reply:
Yeah, I was wondering why we were calling `callback.invokeAfterLoadActions` two times when this flag was set. I'll remove it.